examples/chat/src/network/stream/codec.rs

```diff
@@ -0,0 +1,118 @@
+use bytes::{Bytes, BytesMut};
+use serde::{Deserialize, Serialize};
+use thiserror::Error;
+use tokio_util::codec::{Decoder, Encoder, LengthDelimitedCodec};
+
+#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
+pub struct Message {
+    pub data: Vec<u8>,
+}
+
+#[derive(Debug, Error)]
+#[error("CodecError")]
+pub enum CodecError {
+    StdIo(std::io::Error),
+    EncodeDecode(std::io::Error),
+    SerDe(serde_json::Error),
+}
+
+impl From<std::io::Error> for CodecError {
+    fn from(err: std::io::Error) -> Self {
+        CodecError::StdIo(err)
+    }
+}
+
+#[derive(Debug)]
+pub(crate) struct MessageJsonCodec;
+
+impl Decoder for MessageJsonCodec {
+    type Item = Message;
+    type Error = CodecError;
+
+    fn decode(&mut self, src: &mut BytesMut) -> Result<Option<Self::Item>, Self::Error> {
+        let mut length_codec = LengthDelimitedCodec::new();
+        let frame = match length_codec.decode(src) {
+            Ok(Some(frame)) => frame,
+            Ok(None) => return Ok(None),
+            Err(e) => return Err(CodecError::EncodeDecode(e)),
+        };
```

Comment:
I think you can skip making this it's own variant, merge it with `StdIo`

```suggestion
        let Some(frame) = length_codec.decode(src)? else {
            return Ok(None);
        };
```